cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java

```diff
@@ -772,6 +775,85 @@ public void testListAndGetForPaginatedAPIWithLatestOnlyLCMFlagEnabled() throws E
     setLCMFlag(false);
   }
 
+  @Test
+  public void testListAndGetForPaginatedAPIWithLatestOnlyLCMFlagEnabledForVersionHistory() throws Exception {
+    setLCMFlag(false);
+    Id.Namespace ns1 = Id.Namespace.from(TEST_NAMESPACE1);
+    Id.Artifact ns1ArtifactId = Id.Artifact.from(ns1, AllProgramsApp.class.getSimpleName(), "1.0.0-SNAPSHOT");
+    Id.Application appId = Id.Application.from(ns1, AllProgramsApp.NAME);
+    ApplicationSpecification appSpec = Specifications.from(new AllProgramsApp());
+
+    // Insert one version of the app through AppMetadataStore
+    // This version represents an application before 6.8
+    TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore store = AppMetadataStore.create(context);
+      store.writeApplication(ns1.getId(),
```

Comment:
we can't use this method because it takes a ChangeDetail where the creation time is never null (which is good). Instead, we should be calling the lower level methods in the DefaultStore that directly write the row, passing in null for the creation time.

This test should be moved to DefaultStoreTest instead of in the handler, then you can call package private methods on DefaultStore directly. In general, unit tests should live as close to the relevant logic as possible, testing very focused parts. An edge case like this in the store shouldn't need a test in the handler level, which needs to go through http and the service before getting to the important logic in the store.

Reply:
hmmm, so it's kinda tricky that because `creationTime` is indexed, we cannot use `DefaultStore.addApplication` or `AppMetadataStore.writeApplication` to insert a row with null ChangeDetail. I think we have to add the test inside `AppMetadataStoreTest` instead and expose `getApplicationSpecificationTable()` for testing: this way we could insert the wanted row

Reply:
As offline discussed, will add the test in DefaultStoreTest, adding a null checking when writing the application